crates/macro-support/src/parser.rs

```diff
@@ -803,6 +803,57 @@ pub(crate) fn is_js_keyword(keyword: &str) -> bool {
     JS_KEYWORDS.contains(&keyword)
 }
 
+/// Replace instances of `Self` with the actual type `T`
+fn replace_self(ty: syn::Type, self_ty: Option<&Ident>) -> syn::Type {
+    let self_ty = match self_ty {
+        Some(i) => i,
+        None => return ty,
+    };
+
+    let mut path = match get_ty(&ty) {
+        syn::Type::Path(syn::TypePath { qself: None, path }) => path.clone(),
+        other => return other.clone(),
+    };
+
+    if path.segments.len() == 1 && path.segments[0].ident == "Self" {
+        let new_path = self_ty.clone().into();
+        return syn::Type::Path(syn::TypePath {
+            qself: None,
+            path: new_path,
+        });
+    }
+
+    fn walk_path_change_self(path: &mut syn::Path, self_ty: &Ident) {
+        for seg in path.segments.iter_mut() {
+            if let syn::PathArguments::AngleBracketed(ref mut brackets) = seg.arguments {
+                let types = brackets.args.iter_mut().filter_map(|arg| match arg {
+                    syn::GenericArgument::Type(ref mut ty) => Some(ty),
+                    _ => None,
+                });
+                for ty in types {
+                    match ty {
+                        syn::Type::Path(syn::TypePath { ref mut path, .. }) => {
+                            if path.segments.len() == 1 && path.segments[0].ident == "Self" {
+                                path.segments[0].ident = self_ty.clone();
+                            } else {
+                                walk_path_change_self(path, self_ty)
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+            }
+        }
+    }
+
+    let path_str = quote::quote! { #path }.to_string();
+    if path_str.contains("Self") {
```

Comment:
Couldn't we just call `walk_path_change_self()` and let it just do it's thing anyway? I'm assuming `to_token_stream()` and `to_string()` have to go through the `Path` to build it anyway, so it wouldn't be more expensive.

Reply:
Yeah now that you mentioned it, seems like there is no way around it. Will push an update later today!